Added Auxiliary Circle Feature For Ellipse

iamrajiv opened the pull request:
Auxiliary Circle of an ellipse is the circumcircle of an ellipse i.e. the circle whose centre concurs with that of the ellipse and whose radius is equal to the ellipse's semi-major axis.

The centre of the auxiliary circle is always the same as the ellipse's centre, the circle itself touches both vertices. Hence, the distance between a vertex and the centre is the circle's radius. In an ellipse, the distance between a vertex and the centre is given by the semi-major axis.

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->

#### Other comments

I hope guidance and feedback from fellow mentors.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

- geometry
  - Added Auxiliary Circle Feature For Ellipse

<!-- END RELEASE NOTES -->

iamrajiv committed:
Update ellipse.py

iamrajiv committed:
Update ellipse.py

Yathartha22 commented:
Please add tests and also try more descriptive commit messages, you can refer here: https://github.com/sympy/sympy/wiki/Development-workflow#writing-commit-messages

sidhantnagpal commented:
You do not need to create multiple PRs to address reviews. There have been several PRs for the same #15318, #15281, #15270, #15185. You can push to the respective branch and the changes will appear in the PR automatically.

iamrajiv committed:
Added Test  for Auxiliary Circle of Ellipse

iamrajiv commented:
@sidhantnagpal sorry for inconvenience earlier I did not know about the workflow of sympy and by mistake, I have deleted my branch too because of this there is many PR of the same topic by me.
thank you for help @Yathartha22 @sidhantnagpal

sylee957 commented:
My idea is that this function would better return a geometric object like `Circle(center, radius)` than bare equation.
I think it can make the code more reusable, while still able to access equations via `self.auxiliary_circle().equation()` by reusing the `equation()` method of any circle object.

And also most of the other existing methods seems reusing the geometric objects if possible, other than depicting them with more raw objects.

iamrajiv committed:
updated the code to return geometric object not bare equation

Earlier the code return a bare equation of the auxiliary circle of the ellipse but after the changing the code now it returns the geometric object like in this  case it will return the equation of a circle

iamrajiv committed:
correction in test case for auxiliary circle of ellipse

iamrajiv commented:
thank you @sylee957 for the suggestion

iamrajiv committed:
some correction in code to pass the checks

iamrajiv committed:
fixed Doctests error and failed example error

the code failed to pass the checks because it does not match the return value to expected value

iamrajiv committed:
changes in import

iamrajiv committed:
Update ellipse.py

sylee957 reviewed (commented):
I don't think that `x` and `y` are needed in this implementation

dandiez commented:
Did you consider using ` Max(self.hradius, self.vradius) ` to define the radius (instead of the if blocks)?

iamrajiv committed:
remove x and y because not needed in this implementation

iamrajiv commented:
thank you @sylee957 yes implementation of `x` and `y` is not needed since code run well without it

iamrajiv commented:
@dandiez yes I have used `Max` function locally it gives the correct output but what is the difference both `if` and `Max` gives the same result

iamrajiv committed:
removed trailing whitespace

iamrajiv commented:
@sylee957 using `Max` instead of `if` function is better or not can you give some suggestion on it

dandiez commented:
I guess the advantage of having less lines of code (in this case) is that it is arguably:

-  easier to understand (the function Max is a commonly used function and the reader does not have to figure out how the condition in the "if" statement is related to the return value)
- easier to maintain and test (no need to keep two very similar return statements "synchronised")

iamrajiv commented:
@dandiez its good  thank you for the suggestion

iamrajiv committed:
changed if function to Max function to return the output

used Max function instead of if function because it requires less line of code also it is easier to read and test the code

oscarbenjamin commented on sympy/geometry/ellipse.py:
```diff
@@ -1005,6 +1005,45 @@ def semilatus_rectum(self):
+        x : str, optional
+            Label for the x-axis. Default value is 'x'.
+        y : str, optional
+            Label for the y-axis. Default value is 'y'.
```
These parameters are not there any more so this should be removed from the docstring.

iamrajiv replied:
yup these parameters are not needed at present forget to delete it thank you @oscarbenjamin

oscarbenjamin commented on sympy/geometry/ellipse.py:
```diff
@@ -1005,6 +1005,45 @@ def semilatus_rectum(self):
+        Returns
+        =======
+
+        equation : sympy expression
```
Looks like it actually returns a Circle object. Can you check how other geometry docstrings report their return types and make sure this is consistent.

iamrajiv replied:
when I used the ellipse of the type having both `hradius` and `vradius` different so it gives
```
>>> from sympy import *
>>> e1 = Ellipse(Point2D(1, 5), 2, 3)
>>> e = e1.auxiliary_circle()
>>> e.equation()
(x - 1)**2 + (y - 5)**2 - 9
```
but in case of ellipse having `hradius`  = `vradius`  it gives
```
>>> from sympy import *
>>> e1 = Ellipse(Point2D(1, 5), 2, 2)
>>> e = e1.auxiliary_circle()
>>> e.equation()
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-7-d07ce3f45221> in <module>()
----> 1 e.equation()

AttributeError: 'Add' object has no attribute 'equation'
```
but when I used this
```
>>> from sympy import *
>>> e1 = Ellipse(Point2D(1, 5), 2, 2)
>>> e = e1.auxiliary_circle()
>>> e
(x - 1)**2 + (y - 5)**2 - 4
```
it gives the result but not with `e.equation()` this happen with circle object too i.e.
```
>>> from sympy import *
>>> e1 = Circle(Point2D(1, 5), 4)
>>> e = e1.auxiliary_circle()
>>> e.equation()
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-11-d07ce3f45221> in <module>()
----> 1 e.equation()

AttributeError: 'Add' object has no attribute 'equation'
>>> e
(x - 1)**2 + (y - 5)**2 - 16
```
why this happens I want this code to be consistent means it should return the auxiliary circle equation with single code `self.equation()` for all cases whether it is any ellipse or circle

iamrajiv replied:
ping @oscarbenjamin

oscarbenjamin replied:
I just tested each of those examples with this PR and they all seem to work fine. I didn't see the errors you report and in each the result returns by `e1.auxiliary_circle` was a `Circle` object.

iamrajiv replied:
@oscarbenjamin but `e.equation()` does not work for a circle it does not return the auxiliary circle is it returning for your case?

oscarbenjamin replied:
I don't understand the issue you are describing here. `e.equation` works for a circle as far as I can tell:
```julia
In [1]: Circle((0, 0), 1)                                                                                                                      
Out[1]: Circle(Point2D(0, 0), 1)

In [2]: Circle((0, 0), 1).equation()                                                                                                           
Out[2]: 
 2    2    
x  + y  - 1
```

iamrajiv replied:
I am talking of this whole code, I had run this on jupyter notebook
```
>>> from sympy import *
>>> e1=Circle((0, 0), 1)
>>> e=e1.auxiliary_circle()
>>> e.equation()
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-22-de93a72cf360> in <module>()
      2 e1=Circle((0, 0), 1)
      3 e=e1.auxiliary_circle()
----> 4 e.equation()

AttributeError: 'Add' object has no attribute 'equation'
```
above does not work for circle and ellipse having `hradius=vradius` though both are same 
it only returns auxiliary cirlce when I used below code
```
>>> from sympy import *
>>> e1=Circle((0, 0), 1)
>>> e=e1.auxiliary_circle()
>>> e
x**2 + y**2 - 1
```

iamrajiv replied:
ping @oscarbenjamin

oscarbenjamin replied:
Is it working for you now? I just checked out this PR and tried your code above and it seems to work as expected:
```julia
In [1]: e1=Circle((0, 0), 1)                                                                                                                   

In [2]: e=e1.auxiliary_circle()                                                                                                                

In [3]: e.equation()                                                                                                                           
Out[3]: 
 2    2    
x  + y  - 1

In [4]: e                                                                                                                                      
Out[4]: Circle(Point2D(0, 0), 1)
```

oscarbenjamin replied:
Please don't ping me like this. I receive all email updates anyway. Apparently that ping arrived while I was writing my comment above. If I haven't responded yet then unless it's been at least a week pinging won't make me respond any faster.

iamrajiv replied:
sorry for ping and inconvenience  @oscarbenjamin but no this is not  working on jupyter notebook where are you running this code and how
![111](https://user-images.githubusercontent.com/42106787/53287791-549aad80-37a7-11e9-8c3e-0d5b436f726f.JPG)

oscarbenjamin replied:
I don't think you are running the code from this PR. Close jupyter etc and restart them. Then check the code that is actually in the files to see if it matches what is in this PR.

iamrajiv replied:
Yes both have the same code since I have clone sympy repo in Anaconda 3 also and used the same code what is in PR in ellipse.py file of Anaconda 3 sympy repo and used Anaconda prompt to run jupyter notebook @oscarbenjamin

oscarbenjamin commented on sympy/geometry/ellipse.py:
```diff
@@ -1005,6 +1005,45 @@ def semilatus_rectum(self):
+        Circle(Point2D(2, 5), 4)
+
+        """
+        return Circle(Point2D(self.center.x, self.center.y), Max(self.hradius, self.vradius))
```
Can we just use self.centre instead of `Point2D(self.center.x, self.center.y)`?

iamrajiv replied:
Yes, `self.center` works fine thanks @oscarbenjamin

oscarbenjamin reviewed (commented):
This looks better. I've left some comments

iamrajiv committed:
removed parameters from docstring

iamrajiv committed:
edited return code used self.center instead of self.center.x and self.center.y

smichr commented on sympy/geometry/ellipse.py:
```diff
@@ -1005,6 +1005,37 @@ def semilatus_rectum(self):
         return self.major * (1 - self.eccentricity ** 2)
 
+    def auxiliary_circle(self):
+        """Returns equation of auxiliary circle of the ellipse.
```
This needs to be updated to reflect that it is not returning an equation.

iamrajiv replied:
I have updated the return docstring is it good

iamrajiv replied:
I have updated the return docstring is it good @smichr

oscarbenjamin commented on sympy/geometry/ellipse.py:
```diff
@@ -1005,6 +1005,37 @@ def semilatus_rectum(self):
+        Circle(Point2D(2, 5), 4)
+
+        """
+        return Circle(Point2D(self.center), Max(self.hradius, self.vradius))
```
Is self.center already a Point2D? Can we not just use it directly as in 
```
return Circle(self.center, Max(self.hradius, self.vradius))
```

iamrajiv replied:
Yes, the code works fine with `self.center` I think the class circle has already defined the center of a circle as of dimension 2

iamrajiv replied:
Yes, the code works fine with `self.center` I think the class circle has already defined the center of a circle as of dimension 2 @oscarbenjamin

oscarbenjamin commented:
Do any of the other shapes in the geometry module have analogous methods?

I wonder if there could be a more generic method name for finding circumscribing circles for different shapes. The concept of "auxiliary circle" seems to be specific to ellipses.

smichr commented on sympy/geometry/tests/test_ellipse.py:
```diff
@@ -463,3 +463,39 @@ def test_circumference():
+
+
+def test_auxiliary_circle():
+    x, y = symbols('x, y')
```
The beauty of using Max is that you don't need all the test cases. The following is sufficient:
```python
x, y, a, b = symbols('x y a b')
e = Ellipse((x, y), a, b))
assert e.auxiliary_circle == Circle((x, y), Max(a, b))
```

iamrajiv replied:
thank you @smichr can you review my test cases should I update my test cases with above code or is it good?

smichr commented:
Would there be a problem in simply calling the method `circumcircle`? The definition of auxiliary circle starts by saying that is the circumcircle of the ellipse...so why not just name it as such without one more level of indirection?

iamrajiv committed:
updated return type in  the docstring and removed usage of Point2D in center

used circle object as return type in the docstring instead of the equation and also removed usage of Point2D in center since in the class circle it is already defined the dimension of a center as 2

oscarbenjamin commented:
I'm not familiar with this terminology but Wolfram defines a circumcircle as being specific to a triangle:
[http://mathworld.wolfram.com/Circumcircle.html](http://mathworld.wolfram.com/Circumcircle.html)

smichr committed:
Update ellipse.py

iamrajiv committed:
updated test cases

smichr commented:
Please update tests -- geometry tends to be slow as it is and it would be good to avoid unnecessary tests.

iamrajiv commented:
Yes, I have updated the tests thank you @smichr for guidance

iamrajiv commented:
@smichr is Travis CI not working can you help?

smichr commented:
> can you help?

I just closed and reopened to trigger a restart since the last entry didn't have a hash that agreed with your latest commit. This should be ready once tests pass.

iamrajiv commented:
thank you @smichr

smichr commented:
two errors:
```
________ sympy/geometry/tests/test_ellipse.py:test_object_from_equation ________
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/utilities/runtests.py", line 1306, in _timeout
    function()
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/geometry/tests/test_ellipse.py", line 35, in test_object_from_equation
    assert Circle(x**2 + y**2 + 3*x + 4*y + 26) == Circle(Point2D(-S(3)/2, -2), sqrt(79)*I/2)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/geometry/ellipse.py", line 1426, in __new__
    return Circle((center_x, center_y), sqrt(r2), evaluate=evaluate)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/geometry/ellipse.py", line 1440, in __new__
    r = Point(args[1], 0, evaluate=evaluate).x  # convert via Point as necessary
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/geometry/point.py", line 158, in __new__
    raise ValueError('Imaginary coordinates are not permitted.')
ValueError: Imaginary coordinates are not permitted.
________________________________________________________________________________
____________ sympy/geometry/tests/test_ellipse.py:test_issue_15797 _____________
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/utilities/runtests.py", line 1306, in _timeout
    function()
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/geometry/tests/test_ellipse.py", line 477, in test_issue_15797
    assert u.center.x.is_Float
AssertionError
```

The first could use `Circle(Circle((0, 1), 2).equation()) == Circle((0, 1), 2)`. I am not sure what is wrong with the second one.

smichr committed:
Update test_ellipse.py

smichr committed:
Update test_ellipse.py

iamrajiv commented:
thank you @smichr for guidance and all the other mentors too :+1:
